src/TestFramework/TestFramework/Assertions/Assert.cs

```diff
@@ -40,6 +40,70 @@ internal static void ThrowAssertFailed(string assertionName, string? message)
         => throw new AssertFailedException(
             string.Format(CultureInfo.CurrentCulture, FrameworkMessages.AssertionFailed, assertionName, message));
 
+    /// <summary>
+    /// Helper function that creates and throws an AssertionFailedException with expected and actual values.
+    /// </summary>
+    /// <typeparam name="T">
+    /// The type of the expected and actual values.
+    /// </typeparam>
+    /// <param name="assertionName">
+    /// name of the assertion throwing an exception.
+    /// </param>
+    /// <param name="message">
+    /// The assertion failure message.
+    /// </param>
+    /// <param name="expected">
+    /// Expected value to store in exception data.
+    /// </param>
+    /// <param name="actual">
+    /// Actual value to store in exception data.
+    /// </param>
+    [DoesNotReturn]
+    [StackTraceHidden]
+    internal static void ThrowAssertFailed<T>(string assertionName, string? message, T? expected, T? actual)
+    {
+        AssertFailedException exception = new(
+            string.Format(CultureInfo.CurrentCulture, FrameworkMessages.AssertionFailed, assertionName, message));
+
+        // Store expected and actual values in exception Data for types with known good ToString implementations
+        if (HasKnownGoodToString(expected))
+        {
+            exception.Data["assert.expected"] = expected;
+        }
+
+        if (HasKnownGoodToString(actual))
+        {
+            exception.Data["assert.actual"] = actual;
+        }
+
+        throw exception;
+    }
+
+    private static bool HasKnownGoodToString([NotNullWhen(true)] object? value)
```

Comment:
We could improve the logic to try to detect if the given type overrides `ToString` in which case we call it. We could also allow users to register a method for the display representation of types. 

But all of that can happen later, we are already improving experience as-is.